### PR TITLE
chore(deps): bump aedes-persistence@9.0.3

### PR DESCRIPTION
**Pull request by seriousme**
This PR bumps `aedes-level to 9.0.2`

Kind regards,
Hans

### CODE REVIEW

**Comment by seriousme** [comment]
I also tried to upgrade Tape to 5.5.3, but that fails. 

Reason is that abstract.js sometimes expects a returned packet to be:
```javascript 
Packet { cmd: 'publish', brokerId: 'broker-42', brokerCounter: 43, topic: 'hello', payload: <Buffer 6d 61 74 74 65 6f>, qos: 1, retain: false, dup: false } 
```
But it receives:
```javascript 
{ cmd: 'publish', brokerId: 'broker-42', brokerCounter: 43, topic: 'hello', payload: <Buffer 6d 61 74 74 65 6f>, qos: 1, retain: false, dup: false }
```
(no Class identifier)
Do we want to fix this, and if so how ? Do we want abstract.js to be less strict, of this module to add Class identifiers on each packet ?

Kind regards,
Hans

**Comment by seriousme** [comment]
Just noticed node 16 and 18 failing, I will look into that.

**Comment by robertsLando** [comment]
> Do we want abstract.js to be less strict, of this module to add Class identifiers on each packet ?

I think that it should not be that strict, just compare the packets to see if them are equal without checking the class

**Comment by seriousme** [comment]
Small update:

> Just noticed node 16 and 18 failing, I will look into that.

I did some research and while I managed to fix 1 item, I still found another one and it boils down to a test in `abstract.js` in `aedes-persistence`

We just successfully queued 32 items and verified that there are indeed 32 items in the outgoing queue.
Then this part of the code comes along:

```javascript
  function clearMessage(p) {
      return new Promise((resolve, reject) => {
        instance.outgoingClearMessageId(client, p, (err, received) => {
          if (err) {
            return reject(err)
          }
          resolve(received)
        })
      })
    }

const oStream = instance.outgoingStream(client)
for await (const p of oStream) {
   const received = await clearMessage(p)
   t.deepEqual(received, p, 'must return the packet')
   console.log("deleted:", received.brokerCounter)
}
console.log("Remaining lenght:", oStream.readableLength)
```
I can verify that all 32 items have been read from LevelDB however:

On NodeJS 14 this results in `Remaining lenght: 0`
On NodeJS 16 this results in `Remaining lenght: 15`
On NodeJS 18 this results in: 
```sh
node:internal/errors:465
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_STREAM_PREMATURE_CLOSE]: Premature close
    at new NodeError (node:internal/errors:372:5)
    at Readable.onclose (node:internal/streams/end-of-stream:139:30)
    at Readable.emit (node:events:527:28)
    at emitCloseNT (node:internal/streams/destroy:132:10)
    at process.processTicksAndRejections (node:internal/process/task_queues:81:21) {
  code: 'ERR_STREAM_PREMATURE_CLOSE'
}

Node.js v18.0.0
```
The whole thing goes belly up as soon as `oStream.readableLength` starts to go down.It starts stable at 15 and when it starts to decrease to 14,13,12  etc on NodeJS 14, the whole loop stops at NodeJS 16 and the error is thrown at NodeJS 18.
It must have something to do with  `instance.outgoingClearMessageId` as a loop without it that just counts the items in queue does not have this issue.  It also does not seem to be a race condition on the iterator of `Level`. I tried a small scale test where I loaded 32 items in a leveldb, and iterated over them while deleting items and did not reproduce this behaviour.

Next step will be to see if I can replicate this behaviour on a smaller scale and rule out any relation with `tape` etc..

Kind regards,
Hans

**Comment by seriousme** [comment]
I found the culprit:  :-)

In [abstract.js](https://github.com/seriousme/aedes-persistence/blob/a26e1176525c98403ae9e4e045c5d9b37304521e/abstract.js#L54-L61) I added support for legacy streams to maintain compatibility with `from2` which is not providing Streams3 compatible streams.

This works fine in `aedes-persistence` but fails in `aedes-persistence-level` as it always does a `wrap` on Readables and that causes the issues. Without the wrap the tests succeed at 9.0.2  

I propose to change the condition in `abstract.js` to:
```javascript
   // legacy third party streams are typically not iterable
  function iterableStream (stream) {
    if (opts.legacyStream) {
      return new Readable({ objectMode: true }).wrap(stream)
    }
    return stream
  }
  // end of legacy third party streams support
```
If you agree I'll do another PR on  `aedes-persistence`  with this fix in it and the `just compare the packets to see if them are equal without checking the class` as well,so we can also bump `tape` to 5.x.x

Any persistence modules still providing pre-Streams3 streams will then need to either:
- `wrap` the stream themselves (preferred)
- add option `{ legacyStream: true}` 

to succeed testing against `aedes-persistence` 9.0.x

Kind regards,
Hans

**Comment by robertsLando** [comment]
> I found the culprit: :-)

nice catch @seriousme ! That was not easy to find :)

File: package.json [hunk 1]
```diff
@@ -46,7 +46,7 @@
   "homepage": "https://github.com/mcollina/aedes-persistence-level#readme",
   "devDependencies": {
     "aedes": "^0.46.3",
-    "aedes-persistence": "^8.1.3",
+    "aedes-persistence": "^9.0.2",
```
Comment (robertsLando) [inline package.json]:
```suggestion
    "aedes-persistence": "^9.0.3",
```

**Pull request merged**